pkg/addon/agent/agent.go

```diff
@@ -0,0 +1,276 @@
+package agent
+
+import (
+	"context"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/x509"
+	"encoding/pem"
+	"time"
+
+	"open-cluster-management.io/cluster-proxy/pkg/addon/operator/authentication/selfsigned"
+	proxyv1alpha1 "open-cluster-management.io/cluster-proxy/pkg/apis/proxy/v1alpha1"
+	"open-cluster-management.io/cluster-proxy/pkg/common"
+
+	"open-cluster-management.io/addon-framework/pkg/agent"
+	addonv1alpha1 "open-cluster-management.io/api/addon/v1alpha1"
+	clusterv1 "open-cluster-management.io/api/cluster/v1"
+
+	appsv1 "k8s.io/api/apps/v1"
+	csrv1 "k8s.io/api/certificates/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var _ agent.AgentAddon = &proxyAgent{}
+
+const (
+	ProxyAgentSignerName = "open-cluster-management.io/proxy-agent-signer"
+)
+
+func NewProxyAgent(runtimeClient client.Client, nativeClient kubernetes.Interface, signer selfsigned.SelfSigner) agent.AgentAddon {
+	return &proxyAgent{
+		runtimeClient: runtimeClient,
+		nativeClient:  nativeClient,
+		selfSigner:    signer,
+	}
+}
+
+type proxyAgent struct {
+	runtimeClient client.Reader
+	nativeClient  kubernetes.Interface
+	selfSigner    selfsigned.SelfSigner
+}
+
+func (p *proxyAgent) Manifests(managedCluster *clusterv1.ManagedCluster, addon *addonv1alpha1.ManagedClusterAddOn) ([]runtime.Object, error) {
+	clusterAddon := &addonv1alpha1.ClusterManagementAddOn{}
+	if err := p.runtimeClient.Get(context.TODO(), types.NamespacedName{
+		Name: addon.Name,
+	}, clusterAddon); err != nil {
+		return nil, err
+	}
+	config := &proxyv1alpha1.ManagedProxyConfiguration{}
+	if err := p.runtimeClient.Get(context.TODO(), types.NamespacedName{
+		Name: clusterAddon.Spec.AddOnConfiguration.CRName,
+	}, config); err != nil {
+		return nil, err
+	}
+	lbEndpoint := ""
+	if config.Spec.ProxyServer.Entrypoint.Type == proxyv1alpha1.EntryPointTypeLoadBalancerService {
+		if entrySvc, err := p.nativeClient.CoreV1().
+			Services(config.Spec.ProxyServer.Namespace).
+			Get(context.TODO(), config.Spec.ProxyServer.Entrypoint.LoadBalancerService.Name, metav1.GetOptions{}); err == nil {
+			if len(entrySvc.Status.LoadBalancer.Ingress) > 0 {
+				lbEndpoint = entrySvc.Status.LoadBalancer.Ingress[0].IP
+			}
+		}
+	}
+	deploying := []runtime.Object{
+		newCASecret(addon.Spec.InstallNamespace, AgentCASecretName, p.selfSigner.CAData()),
```

Comment:
we may also need to deploy install namesapce

Reply:
fixed